### PR TITLE
Make Slice.reduce() fully canonical

### PR DESCRIPTION
**Pull request by asmeurer**
This means if two slices are always equal, then `reduce` will reduce them to the same thing (always equal for all array shapes if no shape is passed, or always equal for the given shape for `reduce(shape)`). The updated docstring of `Slice.reduce` details more properties that are true for reduced slices.

Fixes https://github.com/Quansight/ndindex/issues/20

### CODE REVIEW

File: ndindex/slice.py [hunk 1]
```diff
@@ -373,7 +491,10 @@ def as_subindex(self, index):
             return Tuple(self).as_subindex(index)
 
         if isinstance(index, Integer):
-            s = self.as_subindex(Slice(index.args[0], index.args[0] + 1))
+            if index.args[0] == -1:
```
Comment (scopatz) [inline ndindex/slice.py:494]: What is this condition testing?

Comment (asmeurer) [reply ndindex/slice.py:494]: If the integer index is equal to -1.

Comment (scopatz) [reply ndindex/slice.py:494]: So it doesn't need to test for any negative int?

Comment (asmeurer) [reply ndindex/slice.py]: -1 is the only integer where slice(i, i+1) doesn't index one element.

Comment (scopatz) [reply ndindex/slice.py]: Ahh I see
